ocaml/xenopsd/pvs/pvs_setup.ml

```diff
@@ -0,0 +1,603 @@
+              ; p "%s=%d" Const.client_lsb id_lsb
+              ; p "%s=%d" Const.client_msb id_msb
+              ; "udp"
+              ; p "tp_dst=%d" port
+              ; p "actions=resubmit(,%d)" Const.action_table
+              ]
+            ; [
+                p "cookie=%Ld" id
+              ; p "table=%d" Const.port_table
+              ; p "priority=%d" Const.rule_prio
+              ; p "%s=%d" Const.client_lsb id_lsb
+              ; p "%s=%d" Const.client_msb id_msb
+              ; "udp6"
+              ; p "tp_dst=%d" port
+              ; p "actions=resubmit(,%d)" Const.action_table
+              ]
+            ]
+        in
+
+        iter (* over flows below *)
+          (OVS.add_flow ~debug bridge)
+          [
+            [
+              p "cookie=%Ld" id
+            ; p "table=%d" Const.action_table
+            ; p "priority=%d" (Const.rule_prio - 1)
+            ; p "%s=%d" Const.client_lsb id_lsb
+            ; p "%s=%d" Const.client_msb id_msb
+            ; p "%s=%d" Const.dir Const.client2server
+            ; p "actions=%d" proxy_port
+            ]
+          ; [
+              p "cookie=%Ld" id
+            ; p "table=%d" Const.action_table
+            ; p "priority=%d" Const.rule_prio
+            ; p "%s=%d" Const.client_lsb id_lsb
+            ; p "%s=%d" Const.client_msb id_msb
+            ; p "%s=%d" Const.dir Const.proxy2client
+            ; "actions=NORMAL"
+            ]
+          ; [
+              p "cookie=%Ld" id
+            ; p "table=%d" Const.action_table
+            ; p "priority=%d" (Const.rule_prio - 1)
+            ; p "%s=%d" Const.client_lsb id_lsb
+            ; p "%s=%d" Const.client_msb id_msb
+            ; p "%s=%d" Const.dir Const.server2client
+            ; p "actions=%d" proxy_port
+            ]
+          ]
+  in
+
+  (* now add flows *)
+  let* () =
+    iter (* over flows below *)
+      (OVS.add_flow ~debug pvs_bridge)
+      [
+        ["priority=0"; p "table=%d" Const.port_table; "actions=NORMAL"]
+      ; ["priority=0"; p "table=%d" Const.action_table; "actions=NORMAL"]
+      ]
+  in
+  (* add flows for each server 0..n *)
+  let* () = seq 0 (pvs_server_count - 1) (add_pvs_server pvs_bridge) in
+  (* announce that we are ready *)
+  let* () =
+    let path = hotplug // "pvs-rules-active" in
+    match debug with
+    | true ->
+        p "setup complete: %s" path |> print_endline ;
+        ok ()
+    | false ->
+        write path ""
+  in
+  ok ()
+
+let remove dev vif priv hotplug =
+  D.debug "%s: %s %s %s" __FUNCTION__ vif priv hotplug ;
+  let p = Printf.sprintf in
+  let* vif_uuid = read (priv // "vif-uuid") in
+  let* () = pvs_is_running vif_uuid in
+  let* bridge = read (priv // "bridge") in
+  let* pvs_bridge = OVS.pvs_bridge bridge in
+  let path = hotplug // p "%s-ifindex" (Device.to_string dev) in
+  let* str = read path in
+  let* intf =
+    try Scanf.sscanf str "%d" ok
+    with _e -> error "%s: can't parse %s as a number" __FUNCTION__ str
+  in
```

Comment:
Here also it would be good to log the actual bridge and interface etc. E.g. the old script has
```
logger -t "$LOG_TAG" "Removing rules from $PVS_BRIDGE for $PVS_PROXY_INTERFACE $PVS_VM_INTERFACE/$PVS_IF_ID"
```